lib/plausible/ingestion/request.ex

```diff
@@ -61,38 +84,74 @@ defmodule Plausible.Ingestion.Request do
     end
   end
 
-  defp put_request_params(%__MODULE__{} = request, %{} = request_body) do
-    %__MODULE__{
-      request
-      | event_name: request_body["n"] || request_body["name"],
-        referrer: request_body["r"] || request_body["referrer"],
-        screen_width: request_body["w"] || request_body["screen_width"],
-        hash_mode: request_body["h"] || request_body["hashMode"],
-        props: parse_props(request_body)
-    }
-  end
-
-  defp map_domains(%__MODULE__{} = request, %{} = request_body) do
-    domains =
-      if raw = request_body["d"] || request_body["domain"] do
-        raw
-        |> String.split(",")
-        |> Enum.map(&sanitize_hostname/1)
-      else
-        [sanitize_hostname(request.uri)]
-      end
+  defp put_request_params(changeset, %{} = request_body) do
+    Changeset.change(
+      changeset,
+      event_name: request_body["n"] || request_body["name"],
+      referrer: request_body["r"] || request_body["referrer"],
+      screen_width: request_body["w"] || request_body["screen_width"],
+      hash_mode: request_body["h"] || request_body["hashMode"],
+      props: parse_props(request_body)
+    )
+  end
 
-    %__MODULE__{request | domains: domains}
+  defp put_pathname(changeset) do
+    pathname = get_pathname(changeset.changes[:uri], changeset.changes[:hash_mode])
+    Changeset.put_change(changeset, :pathname, pathname)
   end
 
-  defp put_uri(%__MODULE__{} = request, %{} = request_body) do
-    if url = request_body["u"] || request_body["url"] do
-      %__MODULE__{request | uri: URI.parse(url)}
-    else
-      request
+  defp map_domains(changeset, %{} = request_body) do
+    raw = request_body["d"] || request_body["domain"]
+    raw = if is_binary(raw), do: String.trim(raw)
+
+    case raw do
+      "" ->
+        Changeset.add_error(changeset, :domain, "can't be blank")
+
+      raw when is_binary(raw) ->
+        domains =
+          raw
+          |> String.split(",")
+          |> Enum.map(&sanitize_hostname/1)
+
+        Changeset.put_change(changeset, :domains, domains)
+
+      nil ->
+        from_uri = sanitize_hostname(changeset.changes[:uri])
```

Comment:
We don't load any data into this changeset, we start from a clean slate and use `put_change/3` later on. For this reason `changeset.changes` always has all the fields.

Still this clean slate has default values, e.g. timestamp. I don't think default values are included in the changes map. I suggest using [`Ecto.Changeset.get_field/3`](https://hexdocs.pm/ecto/Ecto.Changeset.html#get_field/3) for almost every case. This function looks at the changes and then falls back on the data, which might be useful as we have schema defaults.

```suggestion
        from_uri =
          changeset
          |> Ecto.Changeset.get_field(:uri)
          |> sanitize_hostname()
```

I'd suggest the same for every `changeset.changes` call.

Reply:
good call, will do, cheers

Reply:
3163b68b